### ISSUE TITLE
[autoscaler][azure] example-full.yaml fails to launch due to faulty head node permissions configuration

### ISSUE THREAD

**Issue by XavierGeerinck** [ray-project/ray#14668]
In the following, could it be that this should be a deep update rather than a normal one?

https://github.com/ray-project/ray/blob/6babd1928c6175803500c1db4ca23b621d89a83f/python/ray/autoscaler/_private/commands.py#L586

E.g. when utilizing the Azure autoscaler, we have a "azure_arm_parameters" key that is being set in 

https://github.com/ray-project/ray/blob/6babd1928c6175803500c1db4ca23b621d89a83f/python/ray/autoscaler/_private/azure/config.py#L106

Next to that it is also being set in the YAML definition which will then overwrite the key

As an example, when running my object changes from:

```json
{
  "azure_arm_parameters": {
    "adminUsername": "ubuntu",
    "publicKey": "<MASKED>"
  }
}
```

towards

```json
{
  "azure_arm_parameters": {
    "vmSize": "Standard_D2s_v3",
    "imagePublisher": "microsoft-dsvm",
    "imageOffer": "ubuntu-1804",
    "imageSku": "1804-gen2",
    "imageVersion": "20.07.06"
  }
}
```

Removing the `adminUsername` and `publicKey` which are required in the Azure ARM templates

**Comment by DmitriGekhtman** [comment]
Hi!
Did you encounter problems launching a Ray cluster on Azure?

**Comment by XavierGeerinck** [comment]
Hi @DmitriGekhtman ! This was indeed when launching a cluster on Azure, it failed to create it due to this issue.

**Comment by DmitriGekhtman** [comment]
Could you post the full cluster launching config yaml and the version of Ray used to launch?

**Comment by XavierGeerinck** [comment]
Definitely, the cluster launching config is the one from the example:

**Version:** 2.0.0.dev0

**Cluster Launch Config:**

```yaml
# A unique identifier for the head node and workers of this cluster.
cluster_name: example-cluster

# The maximum number of workers nodes to launch in addition to the head
# node.
max_workers: 2

# The autoscaler will scale up the cluster faster with higher upscaling speed.
# E.g., if the task requires adding more nodes then autoscaler will gradually
# scale up the cluster in chunks of upscaling_speed*currently_running_nodes.
# This number should be > 0.
upscaling_speed: 1.0

# If a node is idle for this many minutes, it will be removed.
idle_timeout_minutes: 5

# Full Example: https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-full.yaml
setup_commands: [ "sudo apt-get update; sudo apt-get install -y libglib2.0-0"]

# Kubernetes resources that need to be configured for the autoscaler to be
# able to manage the Ray cluster. If any of the provided resources don't
# exist, the autoscaler will attempt to create them. If this fails, you may
# not have the required permissions and will have to request them to be
# created by your cluster administrator.
provider:
    type: kubernetes

    # Exposing external IP addresses for ray pods isn't currently supported.
    use_internal_ips: true

    # Namespace to use for all resources created.
    namespace: ray

    # ServiceAccount created by the autoscaler for the head node pod that it
    # runs in. If this field isn't provided, the head pod config below must
    # contain a user-created service account with the proper permissions.
    autoscaler_service_account:
        apiVersion: v1
        kind: ServiceAccount
        metadata:
            name: autoscaler

    # Role created by the autoscaler for the head node pod that it runs in.
    # If this field isn't provided, the role referenced in
    # autoscaler_role_binding must exist and have at least these permissions.
    autoscaler_role:
        kind: Role
        apiVersion: rbac.authorization.k8s.io/v1
        metadata:
            name: autoscaler
        rules:
        - apiGroups: [""]
          resources: ["pods", "pods/status", "pods/exec"]
          verbs: ["get", "watch", "list", "create", "delete", "patch"]

    # RoleBinding created by the autoscaler for the head node pod that it runs
    # in. If this field isn't provided, the head pod config below must contain
    # a user-created service account with the proper permissions.
    autoscaler_role_binding:
        apiVersion: rbac.authorization.k8s.io/v1
        kind: RoleBinding
        metadata:
            name: autoscaler
        subjects:
        - kind: ServiceAccount
          name: autoscaler
        roleRef:
            kind: Role
            name: autoscaler
            apiGroup: rbac.authorization.k8s.io

    services:
      # Service that maps to the head node of the Ray cluster.
      - apiVersion: v1
        kind: Service
        metadata:
            # NOTE: If you're running multiple Ray clusters with services
            # on one Kubernetes cluster, they must have unique service
            # names.
            name: example-cluster-ray-head
        spec:
            # This selector must match the head node pod's selector below.
            selector:
                component: example-cluster-ray-head
            ports:
                - name: client
                  protocol: TCP
                  port: 10001
                  targetPort: 10001
                - name: dashboard
                  protocol: TCP
                  port: 8265
                  targetPort: 8265

# Specify the pod type for the ray head node (as configured below).
head_node_type: head_node
# Specify the allowed pod types for this ray cluster and the resources they provide.
available_node_types:
  worker_node:
    # Minimum number of Ray workers of this Pod type.
    min_workers: 2
    # Maximum number of Ray workers of this Pod type. Takes precedence over min_workers.
    max_workers: 2
    # User-specified custom resources for use by Ray. Object with string keys and integer values.
    # (Ray detects CPU and GPU from pod spec resource requests and limits, so no need to fill those here.)
    resources: {"foo": 1, "bar": 2}
    node_config:
      apiVersion: v1
      kind: Pod
      metadata:
        # Automatically generates a name for the pod with this prefix.
        generateName: example-cluster-ray-worker-
      spec:
        restartPolicy: Never
        volumes:
        - name: dshm
          emptyDir:
            medium: Memory
        tolerations:
        - key: "kubernetes.azure.com/scalesetpriority"
          operator: "Equal"
          value: "spot"
          effect: "NoSchedule"
        containers:
        - name: ray-node
          imagePullPolicy: Always
          image: rayproject/ray:nightly
          command: ["/bin/bash", "-c", "--"]
          args: ["trap : TERM INT; sleep infinity & wait;"]
          # This volume allocates shared memory for Ray to use for its plasma
          # object store. If you do not provide this, Ray will fall back to
          # /tmp which cause slowdowns if is not a shared memory volume.
          volumeMounts:
          - mountPath: /dev/shm
            name: dshm
          resources:
            requests:
              cpu: 1000m
              memory: 512Mi
            limits:
              # The maximum memory that this pod is allowed to use. The
              # limit will be detected by ray and split to use 10% for
              # redis, 30% for the shared memory object store, and the
              # rest for application memory. If this limit is not set and
              # the object store size is not set manually, ray will
              # allocate a very large object store in each pod that may
              # cause problems for other pods.
              memory: 512Mi
  head_node:
    # The minimum number of worker nodes of this type to launch.
    # This number should be >= 0.
    min_workers: 0
    # The maximum number of worker nodes of this type to launch.
    # This takes precedence over min_workers.
    max_workers: 0
    node_config:
      apiVersion: v1
      kind: Pod
      metadata:
        # Automatically generates a name for the pod with this prefix.
        generateName: example-cluster-ray-head-
        # Must match the head node service selector above if a head node
        # service is required.
        labels:
            component: example-cluster-ray-head
      spec:
        # Change this if you altered the autoscaler_service_account above
        # or want to provide your own.
        serviceAccountName: autoscaler

        restartPolicy: Never

        # This volume allocates shared memory for Ray to use for its plasma
        # object store. If you do not provide this, Ray will fall back to
        # /tmp which cause slowdowns if is not a shared memory volume.
        volumes:
        - name: dshm
          emptyDir:
            medium: Memory
        containers:
        - name: ray-node
          imagePullPolicy: Always
          image: rayproject/ray:nightly
          # Do not change this command - it keeps the pod alive until it is
          # explicitly killed.
          command: ["/bin/bash", "-c", "--"]
          args: ['trap : TERM INT; sleep infinity & wait;']
          ports:
          - containerPort: 6379  # Redis port
          - containerPort: 10001  # Used by Ray Client
          - containerPort: 8265  # Used by Ray Dashboard

          # This volume allocates shared memory for Ray to use for its plasma
          # object store. If you do not provide this, Ray will fall back to
          # /tmp which cause slowdowns if is not a shared memory volume.
          volumeMounts:
          - mountPath: /dev/shm
            name: dshm
          resources:
            requests:
              cpu: 1000m
              memory: 512Mi
            limits:
              # The maximum memory that this pod is allowed to use. The
              # limit will be detected by ray and split to use 10% for
              # redis, 30% for the shared memory object store, and the
              # rest for application memory. If this limit is not set and
              # the object store size is not set manually, ray will
              # allocate a very large object store in each pod that may
              # cause problems for other pods.
              memory: 512Mi


# Command to start ray on the head node. You don't need to change this.
# Note dashboard-host is set to 0.0.0.0 so that kubernetes can port forward.
head_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --head --autoscaling-config=~/ray_bootstrap_config.yaml --dashboard-host 0.0.0.0

# Command to start ray on worker nodes. You don't need to change this.
worker_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379
```

**Comment by DmitriGekhtman** [comment]
Ah.. was expecting to see one of the Azure configs in `https://github.com/ray-project/ray/tree/master/python/ray/autoscaler/azure`

When launching this example on a Kubernetes cluster, the `head_node_config` in the line
```head_node_config.update(head_config["node_config"])``` 
should actually be any empty dict before the execution of the line.

What errors/issues did you see when launching with the above config?

**Comment by XavierGeerinck** [comment]
My apologies, I am confusing this with another issued I opened :) 
 
For the Azure setup one I am indeed utilizing the link you provided and the [`example-full.yaml` file in there.](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/azure/example-full.yaml)

When I execute that one then, it states that publicKey and sshUser are missing since they are being removed from the config

**Comment by DmitriGekhtman** [comment]
Got it, thanks!
I'm going to modify the name of this issue for better visibility.

**Comment by AmeerHajAli** [comment]
@XavierGeerinck , how important is this issue to you? are you blocked on this? is there a specific timeline by which you need this?

**Comment by XavierGeerinck** [comment]
Hi Ameer,

It’s not super crucial but I am blocked by it. I however have another issue open regarding the ray cluster purely running on Kubernetes which is more critical.

Kind regards,
Xavier Geerinck

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by AmeerHajAli** [comment]
Can you please provide a link to the k8s issue?

**Comment by DmitriGekhtman** [comment]
It's this one https://github.com/ray-project/ray/issues/14666

**Comment by XavierGeerinck** [comment]
Yes indeed, my apologies for not linking :)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by DmitriGekhtman** [comment]
@XavierGeerinck , could you check if the changes in the above PR resolve the problem for you?

**Comment by XavierGeerinck** [comment]
Hi @DmitriGekhtman I checked into this and applied the fix (as far as I could see it was only the `config.py` ) that got adapted. However, when running this I still get the same issue and the previous parameters were overwritten.

What I added:

```python
    for node_type in config["available_node_types"].values():
        azure_arm_parameters = node_type.setdefault(
            ["node_config"], {}).setdefault("azure_arm_parameters", {})
        azure_arm_parameters["adminUsername"] = ssh_user
        azure_arm_parameters["publicKey"] = public_key
```

Result I get:

```bash
msrestazure.azure_exceptions.CloudError: Azure Error: InvalidTemplate
Message: Deployment template validation failed: 'The value for the template parameter 'adminUsername' at line '1' and column '280' is not provided. Please see https://aka.ms/resource-manager-parameter-files for usage details.'.
```

When adding 2 logging events after the lines shown below, I still see the parameters being overwritten. Did I miss a change here?

https://github.com/ray-project/ray/blob/ee4fb97c44348a6392dd341f79e80818dd030286/python/ray/autoscaler/_private/commands.py#L582

https://github.com/ray-project/ray/blob/ee4fb97c44348a6392dd341f79e80818dd030286/python/ray/autoscaler/_private/commands.py#L592

**Comment by ijrsvt** [comment]
@XavierGeerinck Two things
1. Can you try to run this with `--no-config-cache`
2. Can you add a log after this line https://github.com/ray-project/ray/blob/971855a35376a74bb227155274a2d49a742eac1c/python/ray/autoscaler/_private/commands.py#L237

**Comment by XavierGeerinck** [comment]
@ijrsvt thanks for the `--no-config-cache` tip! I forgot to utilize it, but it seems to be working now! Config 2 is showing the updated map now! Will keep you posted, but I think it's resolved.

**Comment by XavierGeerinck** [comment]
The config issue has been resolved, but I am seeing something else popping up. Should I make a new issue for this one?

```bash
Traceback (most recent call last):
  File "/home/xanrin/.pyenv/versions/3.8.6/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/xanrin/.pyenv/versions/3.8.6/lib/python3.8/site-packages/ray/autoscaler/_private/updater.py", line 129, in run
    self.provider.set_node_tags(
  File "/home/xanrin/.pyenv/versions/3.8.6/lib/python3.8/site-packages/ray/autoscaler/_private/azure/node_provider.py", line 29, in wrapper
    return f(self, *args, **kwargs)
  File "/home/xanrin/.pyenv/versions/3.8.6/lib/python3.8/site-packages/ray/autoscaler/_private/azure/node_provider.py", line 226, in set_node_tags
    self.compute_client.virtual_machines.update(
AttributeError: 'VirtualMachinesOperations' object has no attribute 'update'
  Failed to setup head node.
```

**Comment by ijrsvt** [comment]
> The config issue has been resolved, but I am seeing something else popping up. 

Great to hear!!! 


> Should I make a new issue for this one?

Yes, that would be great (please assign me to it as well :) )
